### PR TITLE
Feature/sort basic ui

### PR DESCRIPTION
**Pull request by adrianjost**
After hours of frustration where I haven't coded a single line, I've decided to copy the prototype from the other edtrio-team and changed it step by step to meet our requirements. A couple hours later now the result can be seen on the screenshots below.
Sadly I had to throw away our old code structure. Feel free to refactor everything here. I haven't had time for that.

# Features
 - Table based UI to write down the question term and the appropriate solution
 - UI can switch to "student-mode"
   - A student can click on the card to flip between the question term and the solution
   - can go to the next question with a separate button
   - can reset the view state to redo all cards.

# Screenshots
image
image
image

### CODE REVIEW

**Comment by tillprochaska** [comment]
I’ll have a closer look into this later today/this evening. Didn’t manage to review earlier, sorry!

File: src/plugins/sorting-task/index.tsx [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  event.preventDefault();
+  appendNewAnswer(editor, node);
+};
+
+const appendNewAnswer = (editor: any, node: any) => {
+  const newTerm = Block.create({
+    type: "sorting-task_term",
+    nodes: List([Text.create("")]),
+  });
+  const newQuestion = Block.create({
```
Comment (usame-algan) [inline src/plugins/sorting-task/index.tsx:30]: Should this be `newSolution`?

File: src/plugins/sorting-task/index.tsx [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    type: "sorting-task_solution",
+    nodes: List([Text.create("")]),
+  });
+
+  const newAnswer = Block.create({
+    type: "sorting-task_question",
+    nodes: List([
+      newTerm,
+      newQuestion,
```
Comment (usame-algan) [inline src/plugins/sorting-task/index.tsx:39]: Same as above

File: src/plugins/sorting-task/index.tsx [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  const newTerm = Block.create({
+    type: "sorting-task_term",
+    nodes: List([Text.create("")]),
+  });
+  const newQuestion = Block.create({
+    type: "sorting-task_solution",
+    nodes: List([Text.create("")]),
+  });
+
+  const newAnswer = Block.create({
```
Comment (usame-algan) [inline src/plugins/sorting-task/index.tsx:35]: `newQuestion`?

File: src/plugins/plus-menu/PlusMenu.tsx [hunk 4]
```diff
@@ -11,12 +11,14 @@ import {
...
   faExternalLinkSquareAlt,
   faImage,
+  faSort,
 } from "@fortawesome/free-solid-svg-icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import {
   onClickCodeButton,
   onClickIframeButton,
   onClickImageButton,
+  onClickSortButton,
```
Comment (tillprochaska) [inline src/plugins/plus-menu/PlusMenu.tsx:21]: I'd try to be consistent with naimg things. As the plugin is now called `sorting-task`, I think we should name that function `onClickSortingTaskButton`.

```suggestion
  onClickSortingTaskButton,
```

File: src/plugins/plus-menu/PlusMenu.tsx [hunk 5]
```diff
@@ -117,6 +119,11 @@ class PlusMenu extends PureComponent<IPlusMenuProps, IPlusMenuState> {
           "Iframe einfügen",
           onClickIframeButton,
         )}
+        {this.renderBlockButton(
+          faSort,
+          "Sortieraufgabe einfügen",
+          onClickSortButton,
```
Comment (tillprochaska) [inline src/plugins/plus-menu/PlusMenu.tsx:125]:
```suggestion
          onClickSortingTaskButton,
```

File: src/plugins/sorting-task/styles.css [hunk 6]
```diff
@@ -0,0 +1,12 @@
+.question-card{
```
Comment (tillprochaska) [inline src/plugins/sorting-task/styles.css:1]:
```suggestion
.question-card {
```

File: src/plugins/sorting-task/styles.css [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  padding: 32px;
+  border: 1px solid lightgray;
+  position: relative;
+}
+.question-card:after{
```
Comment (tillprochaska) [inline src/plugins/sorting-task/styles.css:6]:
```suggestion
.question-card::after {
```

`:after` (with a single colon) has been used in the past. However, in order to distinguish between pseudo-selectors and actual pseudo-elements, a double colon should be used for the latter.

File: src/plugins/sorting-task/index.tsx [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+            }
+          }
+        >
+          Reset
+        </button>
+      </div>
+    ) : (
+      <div>
+        <b>Sortieraufgabe (Lehreransicht)</b>
+        <table {...attributes}>
```
Comment (tillprochaska) [inline src/plugins/sorting-task/index.tsx:107]: Directly nesting `tr`s inside of a `table` element throws an error. While I think the HTML specs do not require wrapping `thead`, `tbody` elements etc., React doesn’t seem to like it.

```html
<thead>
  <tr>
    <th>Begriff</th>
    <th>Musterlösung</th>
  </tr>
</thead>
<tbody>
  {children}
</tbody>
```

File: src/plugins/plus-menu/actions.ts [hunk 9]
```diff
@@ -61,3 +61,28 @@ export const onClickIframeButton = (event: any, editor: Editor) => {
     }),
   );
 };
+
+export const onClickSortButton = (event: any, editor: Editor) => {
```
Comment (tillprochaska) [inline src/plugins/plus-menu/actions.ts:65]:
```suggestion
export const onClickSortingTaskButton = (event: any, editor: Editor) => {
```

**Review by tillprochaska — CHANGES_REQUESTED** [review]
Thanks for bootstrapping the basic elements! 💯I haven’t understood everything, but this is probably because I’ve not understood all the Slate-specific concepts. However, I’ve added some minor suggestions/comments, but these just details. :)

Also, the current code throws some errors. In the cases I’ve been able to find the cause for them, I’ve added a comment at the relevant lines.

**Comment by tillprochaska** [comment]
In the following, I’ve written down some general thoughts on another possible approach to this task. But that shouldn’t stop us from merging this PR! 👍 

I’m not sure if we should rely that much on Slate.js for the sorting task itself. Building a vanilla React component that doesn’t rely on Slate-specific concepts (like blocks etc.) might make things much easier. Here are some of the reasons:

Slate is great for »freeform content«. That’s why it’s a good choice for the editor and the document itself. On the other hand, the UI for our sorting task component are known upfront and their structure is more or less fixed.

For example, when entering the terms, there should always be two inputs – one for the term and one for the explanation. Using Slate to build this interface will introduce some problems. E. g.,  currently the user is able to delete these two inputs by simply pressing the backspace key [1]. Also, it's currently possible to add code blocks in those cells.

We could probably work around all these issues. But there’s a solutions that’s way more straight-forward. Using basic inputs etc. to compose the interface might just be enough. 

We might even build the interface completely as a stand-alone component and integrate it with the editor at the end of our process. This way, we would not directly depend on the `edtr.io` project or Slate which might be a big advantage as the SC team seems to have some major refactorings (which in turn might result in breaking changes) in the pipeline.

Hopefully that was understandable…?!

***

[1] Here’s a short screencast that demonstrates the bug:

bug

**Review by tom2022 — APPROVED** [review]
As just agreed, approval from my side, let's merge and proceed with the current status.

**Pull request merged**